Allow resizing of parametrized tensors

kazhou opened the pull request:
Summary: Modify `parametrize.py` to allow resizing of parametrized tensors

Test Plan: TODO

Differential Revision: D29279442

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -66,17 +66,20 @@ class ParametrizationList(ModuleList):
     Args:
         modules (iterable): an iterable of modules representing the parametrizations
         original (Parameter or Tensor): parameter or buffer that is parametrized
+        allow_resize (bool): a boolean flag that denotes whether the parametrized tensor can be resized
```
I think we want to be a little more ominous here. Warning that the current Module might not work any-more and users enable this flag at their own risk.

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -66,17 +66,20 @@ class ParametrizationList(ModuleList):
 
     def __init__(
-        self, modules: Iterable[Module], original: Union[Tensor, Parameter]
+        self, modules: Iterable[Module], original: Union[Tensor, Parameter], allow_resize: bool = False
```
nit you can make this a kwarg only argument by adding a `, *, allow_resize` etc

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -231,6 +234,7 @@ def right_inverse(self, X: Tensor) -> Tensor
         tensor_name (str): name of the parameter or buffer on which to register
             the parametrization
         parametrization (nn.Module): the parametrization to register
+        allow_resize (bool): a boolean flag that denotes whether the parametrized tensor can be resized
```
Same comment about the doc

albanD reviewed (commented):
Can you please update the tests in `test/test_nn.py` (and in particular the ones that make sure that an error is raised when the flag is not provided) to make sure this is working as expected.

cc @Lezcano if you have any opinion on the name of the flag.

Lezcano commented:
I would wait first for https://github.com/pytorch/pytorch/pull/58488 to be landed, as that PR heavily extends the current scope of parametrizations.

In that PR we support changing the size of `original`, but not the size of the returned `weight`. In other words, `right_inverse` may change the size of the parameter, but then forward should map this tensor with a new size to a tensor of the old size.

Now, we do a number of checks to protect the user. A number of these checks could be relaxed. Perhaps we could add an `unsafe` kw-only flag and document that, if you use this flag, you may be allowed to do more things, but if you find an error you are on your own.

I'd be happy to submit this after we land https://github.com/pytorch/pytorch/pull/58488

albanD commented:
I'm afraid this one will not be covered by https://github.com/pytorch/pytorch/pull/60530 indeed.
The use case here is to use parametrization in conjunction with a post forward hook to be able to do less computation during the forward of this Module while still returning a full size output.
This will indeed not work for all Module and is pretty unsafe in general, but sounds ok for this particular use case.

albanD commented on test/test_nn.py:
```diff
@@ -2157,6 +2157,11 @@ def forward(self, X):
 
+        class Resize(nn.Module):
+            def forward(self, X):
+                print(X.shape)
```
debug print

albanD commented on test/test_nn.py:
```diff
@@ -2171,6 +2176,41 @@ def forward(self, x):
 
+        # Test allow_resize flag
+        parametrize.register_parametrization(model, "weight", Resize())  # default allow_resize = False
+        self.assertRaises(RuntimeError)
```
I don't think this is doing what you want.

You can use
```python
with self.assertRaisesRegex(RuntimeError, "Your error msg"):
    the_code_that_should_fail()
```
To make sure you get the error you expect.

albanD commented on torch/nn/utils/parametrize.py:
```diff
@@ -66,17 +66,21 @@ class ParametrizationList(ModuleList):
         modules (iterable): an iterable of modules representing the parametrizations
         original (Parameter or Tensor): parameter or buffer that is parametrized
+        allow_resize (bool): a boolean flag that denotes whether the parametrized tensor can be resized,
+            Warning: the current Module might not work anymore, enable this flag at the user's own risk.
```
The person reading this is the user :D 
So, "enable this flag at your own risk"

albanD reviewed (approved):
Looks good, thanks for the update

Lezcano commented on test/test_nn.py:
```diff
@@ -2157,6 +2157,14 @@ def forward(self, X):
 
+        class Resize(nn.Module):
+            def forward(self, X):
+                return X[[0]]
```
What is this notation? I had never seen it before and a quick search does not throw anything meaningful.

kazhou replied:
`X[[0]]` just selects the indices in the list inside the outside brackets i.e. the 0-th row

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -231,6 +235,8 @@ def right_inverse(self, X: Tensor) -> Tensor
         tensor_name (str): name of the parameter or buffer on which to register
             the parametrization
         parametrization (nn.Module): the parametrization to register
+        allow_resize (bool): a boolean flag that denotes whether the parametrized tensor can be resized,
```
```suggestion

Keyword args:
        allow_resize (bool): a boolean flag that denotes whether the parametrized tensor can be resized. Default: `False`
```

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -263,6 +269,8 @@ def right_inverse(self, X: Tensor) -> Tensor
         # Just add the new parametrization to the parametrization list
         module.parametrizations[tensor_name].append(parametrization)  # type: ignore[index, union-attr]
+        # If allow_resize was True in previous parametrization, keep it enabled
+        module.parametrizations[tensor_name].allow_resize |= allow_resize  # type: ignore[index, union-attr]
```
```suggestion
        module.parametrizations[tensor_name].allow_resize = allow_resize  # type: ignore[index, union-attr]
```
as the `allow_resize` parameter just affects the output of the last registered parametrisation. This way the mechanism is always opt-in / opt-out.

z-a-f replied:
The problem with overriding the `allow_resize` is that if you register multiple, when only the early ones are resizing, it will break it. Consider an example

```
register_parametrization(model.linear, 'weight', ResizingParametrization, allow_resize=True)
register_parametrization(model.linear, 'weight', NonResizingParametrization)
```

In the example above, the `allow_resize` is set to `False`, but because of the sequential nature, the first parametrization will resize the tensor, and the whole thing will throw an error.

I think the behavior should be that once you allow resizing, the whole parametrization list becomes unsafe.

Lezcano replied:
Ah, I see what you mean, yes.

It also makes more sense to leave it the way it is now when if we make this into an `unsafe` check. Once you do not check the correctness of one of the parametrisations, you break the invariants and these may not hold from there on.

This point should be documented though.

kazhou replied:
Hi @Lezcano, what do you mean by making this into an `unsafe` check/renaming this flag to `unsafe`? Do you mean like renaming `allow_resize` to `unsafe` and updating the documentation to be more generic e.g. "if enabled, parametrization invariants may not hold anymore"?

Lezcano replied:
Yes, rename `allow_resize` to `unsafe` and turn off quite a few other checks. In that PR mentioned above, we perform quite a few checks both in `ParametrizationList.__init__` and in `register_parametrization`.

In particular, in the file
https://github.com/pytorch/pytorch/blob/f675a3c020d638b64c988457e715c2e561667cb8/torch/nn/utils/parametrize.py
Lines 455-493, should be move them under an `if not unsafe:`.
Same with L169-188 (and also remove the `not self.allow_resize` in L183).

We then should modify the messages raised by these exceptions to announce to the user that this `unsafe` flag exists.

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -108,7 +112,7 @@ def forward(self) -> Tensor:
-        if x.size() != self.original.size():
+        if not self.allow_resize and x.size() != self.original.size():
             raise RuntimeError(
                 "The parametrization may not change the size of the parametrized tensor. "
```
You might want to update this error to point the user to the existence of this flag to make it discoverable.

lezcano reviewed (commented):
This looks good to me, just left a few comments.

As mentioned before, https://github.com/pytorch/pytorch/pull/60530, which should land soon, adds a number of other checks to the parametrizations to make them more usable now that they are becoming slightly more complex. Some of these tests incur in a runtime penalty when calling `register_parametrization`.

I would suggest to wait until https://github.com/pytorch/pytorch/pull/60530 has landed, and then rename this flag to `unsafe` and use it to deactivate then all the checks that we perform.

albanD commented:
@Lezcano this is now rebased on top of the other PR, you can take a second look?

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -162,26 +166,27 @@ def __init__(
+                )
+            if Z.dtype != original_dtype:
+                raise ValueError(
+                    "Registering a parametrization may not change the dtype of the tensor.\n"
```
unless the `unsafe` flag is enabled

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -389,6 +394,9 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
             the parametrization
         parametrization (nn.Module): the parametrization to register
+    Keyword args:
+        unsafe (bool): a boolean flag that denotes whether the parametrization breaks the invariants. Default: `False`
```
Perhaps it's better to be more explicit, as we do not refer to these as "invariants" in the documentation?
> whether the parametrization may change the dtype and shape of the tensor.

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -389,6 +394,9 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
         parametrization (nn.Module): the parametrization to register
+    Keyword args:
+        unsafe (bool): a boolean flag that denotes whether the parametrization breaks the invariants. Default: `False`
+            Warning: the current Module might not work anymore, enable this flag at your own risk.
```
I think that the biggest problem of activating this flag is that the errors that the parametrization may have are not detected early and the errors messages that one gets are notably more cryptic. So perhaps write
> Warning: the parametrization is not checked for consistency upon registration, enable this flag at your own risk

Lezcano replied:
I just realised that we also need to update the `.. note:` in L377-378 reflecting this. A possible rewrite could be:
> If ``unsafe=False`` (default) both the `forward` and `right_inverse` methods will be called once to perform a number of consistency checks.
If ``unsafe=True``, then `right_inverse` will be called if the tensor is not parametrized, and nothing will be called otherwise.

Lezcano replied:
@kazhou you marked these as resolved but perhaps forgot to push the changes?

kazhou replied:
it should be synced from phabricator to this PR? it was re-exported about 2 hrs ago

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -438,56 +446,58 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
                 raise ValueError(
-                    "The tensor returned by parametrization.right_inverse must have the same dtype "
-                    f"as module.{tensor_name}.\n"
+                    "Registering a parametrization may not change the dtype of the tensor.\n"
```
unless the `unsafe` flag is enabled.

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -438,56 +446,58 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
                 raise ValueError(
-                    "The tensor returned by parametrization.right_inverse must have the same shape "
-                    f"as module.{tensor_name}.\n"
+                    "Registering a parametrization may not change the shape of the tensor.\n"
```
unless the `unsafe`  flag is enabled.

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -438,56 +446,58 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
+                    )
+                if Z.dtype != Y.dtype:
+                    raise ValueError(
+                        "The tensor returned by parametrization.right_inverse must have the same dtype "
```
unless the `unsafe` flag is enabled.

Lezcano commented on torch/nn/utils/parametrize.py:
```diff
@@ -438,56 +446,58 @@ def right_inverse(self, X: Tensor) -> Union[Tensor, Sequence[Tensor]]
+                    )
+                if Z.shape != Y.shape:
+                    raise ValueError(
+                        "The tensor returned by parametrization.right_inverse must have the same shape "
```
unless the `unsafe` flag is enabled.

lezcano reviewed (approved):
I left a few suggestions regarding the wording of the documentation, but the functionality looks good to me, thanks!

kazhou committed:
Allow resizing of parametrized tensors (#60418)

Summary:
Modify `parametrize.py` to allow resizing of parametrized tensors

Pull Request resolved: https://github.com/pytorch/pytorch/pull/60418

Test Plan:
`buck test mode/dev-nosan //caffe2/test:nn -- --exact 'caffe2/test:nn - test_register_and_remove_parametrization (test_nn.TestNN)'`

https://pxl.cl/1L0wh

Differential Revision: D29279442

Pulled By: kazhou

fbshipit-source-id: 96c2c5a6b40344710f91413d70c3b0439e3b89e2
